### PR TITLE
fix(compiler-cli): resolve rootDir to absolute

### PR DESCRIPTION
**Pull request by Toxicable**
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
- [x] Feature

## What is the current behavior?
Issue Number: #36290

## What is the new behavior?
Allows for a relative rootDir to be passed, usually via the command line which we will join to the cwd

## Does this PR introduce a breaking change?
- [x] No

### CODE REVIEW

**Comment by Toxicable** [comment]
Im not 100% certain this is the place to make this change so would appciate some feedback on that.

For context I'd like to land this so that we can remove the patch found here: https://github.com/bazelbuild/rules_nodejs/pull/1619/files#diff-5eb8729ae1c44d09b4ddc25c90585b9a

**Review by JoostK — CHANGES_REQUESTED** [review]
I was curious how this is handled in tsc, and they [preprocess the command line args using the internal `ts.convertToOptionsWithAbsolutePaths`](https://github.com/microsoft/TypeScript/blob/2c08affa0d0d7fc55f15ab22e0326b93326d21d8/src/compiler/commandLineParser.ts#L2182-L2214). Maybe we can implement a similar approach, to achieve consistent handling of compiler options.

Regardless, this change would need a test. You can probably extend `NgtscTestEnvironment` with a method to configure additional command line args, and then pass those through in `driveMain` and `driveDiagnostics`.

**Comment by jbedard** [comment]
@Toxicable any update on this? I tried patching this along with the recommendation from @JoostK and it seems to be working. I could take it over if you'd like...

**Comment by Toxicable** [comment]
@jbedard Ah yeah if you wanna take this one then that's cool with me

**Comment by jbedard** [comment]
@JoostK is there an example of a test similar to what you're suggesting with `NgtscTestEnvironment`? Would this be testing a full run of ngtsc as opposed to testing only `readConfiguration` or `readCommandLineAndConfiguration`?

If we were to do something like `convertToOptionsWithAbsolutePaths` in typescript do you think that would go somewhere such as `readConfiguration`?

File: packages/compiler-cli/src/ngtsc/util/src/typescript.ts [hunk 1]
```diff
@@ -93,19 +93,24 @@ export function isExported(node: ts.Declaration): boolean {
...
   }
 
   // In Windows the above might not always return posix separated paths
   // See:
   // https://github.com/Microsoft/TypeScript/blob/3f7357d37f66c842d70d835bc925ec2a873ecfec/src/compiler/sys.ts#L650
   // Also compiler options might be set via an API which doesn't normalize paths
-  return rootDirs.map(rootDir => absoluteFrom(rootDir));
+  return rootDirs.map(rootDir => {
+    const rooted = fs.isRooted(rootDir) ? rootDir : fs.join(cwd, rootDir);
+    return absoluteFrom(rooted);
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/util/src/typescript.ts:112]: Actually you could replace the whole return statement with:

```ts
return rootDirs.map(rootDir => fs.resolve(cwd, rootDir);
```

**Comment by petebacondarwin** [comment]
@jbedard - you could certainly create a simply unit test in a new file `typescript_spec.ts` at https://github.com/angular/angular/tree/master/packages/compiler-cli/src/ngtsc/util/test.

I would run this test using `runInEachFileSystem()` to provide a mock `FileSystem` for the tests.

I am not convinced that an e2e test is particularly needed.

**Comment by literalpie** [comment]
I'm taking a stab at this in #41359

**Comment by petebacondarwin** [comment]
Closing in favour of #41359

**Pull request closed without merging**